README.md
<p align="center">
  <a href="https://encipherr.pythonanywhere.com/">
    <img src="https://github.com/Oussama1403/Encipherr/blob/main/src/static/pwa/512x512-blue.png" width="100">
  </a>
  <h3 align="center">Encipherr</h3>

 
  <p align="center">
    Free online encryption and decryption tool.
    <br>
    <a href="https://encipherr.pythonanywhere.com/"><strong> Explore Encipherr »</strong></a>
  </p>
</p>

<!-- PROJECT SHIELDS -->
<!--
*** I'm using markdown "reference style" links for readability.
*** Reference links are enclosed in brackets [ ] instead of parentheses ( ).
*** See the bottom of this document for the declaration of the reference variables
*** for contributors-url, forks-url, etc. This is an optional, concise syntax you may use.
*** https://www.markdownguide.org/basic-syntax/#reference-style-links
-->
<div align="center">

[contributors-shield]: https://img.shields.io/github/contributors/Oussama1403/Encipherr.svg?style=for-the-badge
[contributors-url]: https://github.com/Oussama1403/Encipherr/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/Oussama1403/Encipherr.svg?style=for-the-badge
[forks-url]: https://github.com/Oussama1403/Encipherr/network/members
[stars-shield]: https://img.shields.io/github/stars/Oussama1403/Encipherr.svg?style=for-the-badge
[stars-url]: https://github.com/Oussama1403/Encipherr/stargazers
[issues-shield]: https://img.shields.io/github/issues/Oussama1403/Encipherr.svg?style=for-the-badge
[issues-url]: https://github.com/Oussama1403/Encipherr/issues
[license-shield]: https://img.shields.io/github/license/Oussama1403/Encipherr.svg?style=for-the-badge
[license-url]: https://github.com/Oussama1403/Encipherr/blob/main/LICENSE.txt

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]


### Made with love using:
![JavaScript](https://img.shields.io/badge/javascript-%23323330.svg?style=for-the-badge&logo=javascript&logoColor=%23F7DF1E) ![HTML5](https://img.shields.io/badge/html5-%23E34F26.svg?style=for-the-badge&logo=html5&logoColor=white) ![CSS3](https://img.shields.io/badge/css3-%231572B6.svg?style=for-the-badge&logo=css3&logoColor=white) ![Python](https://img.shields.io/badge/python-%2314354C.svg?style=for-the-badge&logo=python&logoColor=white) ![Bootstrap](https://img.shields.io/badge/bootstrap-%23563D7C.svg?style=for-the-badge&logo=bootstrap&logoColor=white) ![Flask](https://img.shields.io/badge/flask-%23000.svg?style=for-the-badge&logo=flask&logoColor=white)

</div>

## About The Project

<!-- [![screenshot](screenshot.png)](https://encipherr.pythonanywhere.com/) -->
<a href="https://encipherr.pythonanywhere.com/"><img src="images/scrn_1.png" width="200" />  <img src="images/scrn_2.png" width="200" /></a>

Encipherr is a web app that provides powerful encryption of data. It is fast and free, and it is open source. It uses no ads, and it is the only easy way to encrypt your data in any web browser.

Encipherr uses <a href="https://fr.wikipedia.org/wiki/Advanced_Encryption_Standard" target="_blank">AES</a> encryption algorithm. 
AES is both fast, and cryptographically strong. It is a good default choice for encryption and it's considered one of the strongest algorithms available.  

## Usage
Visit <a href="https://Encipherr.pythonanywhere.com/" target="_blank">Encipherr website</a> \
or install project files and run it locally:

```bash
git clone https://github.com/Oussama1403/Encipherr

```
The project directory will contain:

```
.
├── CHANGELOG.md
├── LICENSE.txt
├── main.py
├── README.md
├── requirements.txt
├── screenshot.png
└── src
    ├── app.py
    ├── config.py
    ├── db.sqlite3
    ├── modules.py
    ├── robots.txt
    ├── routes.py
    ├── static
    │   ├── assets
    │   │   └── logo.png
    │   ├── css
    │   │   └── style.css
    │   ├── easter-egg
    │   │   ├── egg.html
    │   │   └── egg.js
    │   ├── js
    │   │   ├── ajaxcall.js
    │   │   └── upload.js
    │   ├── pwa
    │   │   ├── 512x512-blue.png
    │   │   ├── 512x512.png
    │   │   ├── app.js
    │   │   ├── manifest.json
    │   │   └── offline.html
    │   └── uploads
    ├── sw.js
    └── templates
        ├── about.html
        ├── base.html
        ├── home.html
        ├── page-404.html
        ├── page-500.html
        └── privacy.html

9 directories, 30 files

```

:warning: Make sure python3 and flask framework are installed.

install modules:
```bash
pip install -r requirements.txt
```
then run the server by typing in your terminal:

```python
python3 main.py
```
## CHANGELOG
Read the latest notable changes made to a Encipherr [here](CHANGELOG.md)

## TODO
- Custom encryption key (password) (Done) :heavy_check_mark:.
- More encryption algorithms.
- Offline usage (semi-ready).

## Encipherr-CLI :sparkles:

<b>A command line interface (CLI) version of Encipherr for the offline usage.
<a href="https://github.com/Oussama1403/Encipherr-CLI">github.com/Oussama1403/Encipherr-CLI</a></b>

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to test your changes.

## License
[MIT](https://choosealicense.com/licenses/mit/)


CHANGELOG.md
# Change Log
## 2022-09-04
- Added an option to encrypt with a <b>custom password</b> of the user choice.
## 2022-08-31
- Improve app SEO
## 2022-08-28
- Added custom 404 & 500 error handlers.
- Added 404 & 500 error pages.
- About page redesigned.
## 2022-08-27
- Restructure the project from Single module structure to Package structure (necessary improvements to allow the project to grow or scale.)
- Cleaner code to ease writing, reading and maintaining the project.
- few other bugs fixed.
## 2022-08-25

- Update about section.

## 2022-08-24

- Organize project files: main project files moved to project-src/
- README Updated: Added a Screenshot, Updated project SHIELDS, Update Project Tree
- Added LICENSE file (MIT)
- Added Requirements File

## 2022-08-20

- Update to work with Flask 2.2.2

## 2022-08-08

- Code imporvements

## 2022-04-14

- Small update to the ui.

## 2022-03-23

- All releases deleted for a new version numbering scheme.

## 2022-03-19

- Offline page redesigned.
- Minor changes in the frontend.

## [released] 2021-09-03
- Added an offline page when there is no internet connection.
- Update the service worker for offline page usage.
- Minor changes to the frontend.

## [released] 2021-08-31

- Encipherr is now a progressive web app.
- few changes in the frontend; navbar hover style changed and position is set to fixed-top.
- set 1px letter-spacing to navbar links.
- intro phrase changed.



## [released] 2021-08-23

- Change app name from bsencryptor to encipherr
- Small changes in the ui
- Add title icon
- the old website will automatically redirect you to the new one.



## [released] 2021-08-16
 
### Added
- Added guest Sessions when user enters the website.keeping the user actions secure.
- Session are not permanent.
- Create a private temporary folder when user uploads a file.Will be deleted after 
serving.
- If any error happens when doing operation to the file,will be automatically deleted.


 
### Changed
- Css changes in file encryption section.

 
### Fixed
- The upload folder where files kept temporary for encrypting/decrypting was public and sometimes the files are not deleted which causes a huge security flaw.


src/routes.py
""" App Routing """

from flask import flash,request,render_template,url_for,redirect,send_from_directory,abort,after_this_request,session
from .app import app
from src.modules import TextEncryption,FileEncryption,Utils
import os,shutil

@app.route('/genkey',methods=['GET'])
def genkey():
    return Utils.genkey()

@app.route('/text',methods=['POST'])
def text_mode():
    data = request.get_json() 

    text = TextEncryption()
    
    if data["submit_b"] == "Encrypt Text":
        response = text.encrypt()
        return response
    else:
        response = text.decrypt()
        return response 


@app.route('/',methods=['GET'])
def base():
    return redirect(url_for('home'))

@app.route('/home',methods=['POST','GET'])
def home():
    """Handle all incoming post/get requests"""
    
    if request.method == 'POST':
        Utils.SetupGuestSession()
        if request.form["submit_b"] == "Upload and Encrypt":
            try:
                Utils.Upload_file()
                try:
                    file = FileEncryption()
                    filename=file.encrypt()
                except:
                    flash('Encryption failed! , possible problem: key not found or invalid key')
                    path = session.get('path','not set')
                    shutil.rmtree(path)
                    return redirect(url_for('home'))  
                
                return redirect(url_for('getfile',file_name=filename))
            except:
                flash('Upload failed! , possible problem: no file to upload or key not found')
                return redirect(url_for('home'))

        elif request.form["submit_b"] == "Upload and Decrypt":
            try:
                Utils.Upload_file()
                
                try:
                    file = FileEncryption()
                    filename=file.decrypt()
                except:
                    flash('Decryption failed! , possible problem: key not found or invalid key')
                    path = session.get('path','not set') 
                    shutil.rmtree(path)
                    return redirect(url_for('home'))
                
                return redirect(url_for('getfile',file_name=filename))
            
            except:
                flash('Upload failed! , possible problem: no file to upload or key not found')
                return render_template('home.html')
        else:
            return render_template('home.html')

    else:
        return render_template('home.html')

@app.route("/get-file/<file_name>")
def getfile(file_name):
    """Return file for downloading,after download it will be deleted with the directory"""
    path = session.get('path','not set')
    #print(path)
    filename = session.get('filename','not set')
    try:
        @after_this_request
        def remove_file_and_dir(response):
            if os.path.exists(path):
                if app.config["ENV"] == "DEV":
                    shutil.rmtree(path)
                else:
                    # Delete only the file without its dir to avoid 
                    # OSError: [Errno 16] Device or resource busy: '.nfs0000000005..' prod error.
                    os.remove(os.path.join(path,filename))
            session.pop('path')
            session.pop('filename')
            
            return response

        return send_from_directory(directory=path, path=file_name,as_attachment=True,max_age=0)
    except FileNotFoundError:
        abort(404)

@app.route('/sw.js',methods=["GET","POST"])
def service_worker():
    from flask import make_response
    response = make_response(send_from_directory('.',path='sw.js'))
    response.headers['Content-Type'] = 'application/javascript'
    response.headers['Service-Worker-Allowed'] = '/'
    return response

@app.route('/robots.txt',methods=["GET"])
def robots_file():
    from flask import make_response
    response = make_response(send_from_directory('.',path='robots.txt'))
    response.headers["Content-type"] = "text/plain"
    return response
    
@app.route("/about")
def about():
    return render_template('about.html')

@app.route("/privacy")
def privacy():
    return render_template('privacy.html')

@app.errorhandler(404)
def page_not_found(error):
    return render_template("page-404.html"), 404

@app.errorhandler(500)
def server_error(error):
    return render_template("page-500.html"), 500
